src/types/osstr.rs

```diff
@@ -0,0 +1,191 @@
+use crate::types::PyString;
+#[cfg(not(windows))]
+use crate::PyNativeType;
+use crate::{
+    ffi, AsPyPointer, FromPyObject, IntoPy, PyAny, PyObject, PyResult, PyTryFrom, Python,
+    ToPyObject,
+};
+use std::borrow::Cow;
+use std::ffi::{OsStr, OsString};
+#[cfg(not(windows))]
+use std::os::raw::c_char;
+
+impl ToPyObject for OsStr {
+    fn to_object(&self, py: Python) -> PyObject {
+        // If the string is UTF-8, take the quick and easy shortcut
+        if let Some(valid_utf8_path) = self.to_str() {
+            return PyString::new(py, valid_utf8_path).into();
+        }
+
+        // Take advantage of Python's surrogateescape encoding to support arbitrary byte
+        // sequences instead of limiting ourselves to UTF-8
+
+        // All targets besides windows support the std::os::unix::ffi::OsStrExt API:
+        // https://doc.rust-lang.org/src/std/sys_common/mod.rs.html#59
+        #[cfg(not(windows))]
+        {
+            let bytes = std::os::unix::ffi::OsStrExt::as_bytes(self);
+
+            let ptr = bytes.as_ptr() as *const c_char;
+            let len = bytes.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is this the appropriate function? Does it use the surrogateescape error
+                // handler? Should it be PyUnicode_DecodeLocaleAndSize instead maybe?
+                let pystring = ffi::PyUnicode_DecodeFSDefaultAndSize(ptr, len);
+                py.from_owned_ptr::<PyString>(pystring).into()
+            }
+        }
+
+        #[cfg(windows)]
+        {
+            let wstr: Vec<u16> = std::os::windows::ffi::OsStrExt::encode_wide(self).collect();
+
+            let ptr = wstr.as_ptr();
+            let len = wstr.len() as ffi::Py_ssize_t;
+            unsafe {
+                // HELP: is there some better thing to do here than panic? Some last resort thing to
+                // try?
+                py.from_owned_ptr_or_err::<PyString>(ffi::PyUnicode_FromWideChar(ptr, len))
+                    .expect("OsStr conversion to Python string via FromWideChar failed")
+                    .into()
+            }
+        }
+    }
+}
+
+// It's not possible to implement FromPyObject for OsStr on Windows, because Python's wchar
+// representation can't be directly mapped to OsStr's one
+// HELP: should this implementation be here at all? It may be confusing for PyO3 users when a trait
+// is implemented only on some platforms
+#[cfg(not(windows))]
+impl<'a> FromPyObject<'a> for &'a OsStr {
+    fn extract(ob: &'a PyAny) -> PyResult<Self> {
+        let pystring = <PyString as PyTryFrom>::try_from(ob)?; // Cast PyAny to PyString
+
+        let fs_encoded_bytes: &crate::types::PyBytes = unsafe {
+            ob.py()
+                .from_borrowed_ptr(ffi::PyUnicode_EncodeFSDefault(pystring.as_ptr()))
+        };
+
+        Ok(std::os::unix::ffi::OsStrExt::from_bytes(
+            fs_encoded_bytes.as_bytes(),
+        ))
+    }
+}
```

Comment:
Hmm yeah I agree it's undesirable to have this trait only work on some platforms. A lot of users turn to Rust for the easy multiple-platform support.

Also, we're in the process of designing a rewrite of PyO3's object system which would result in better performance and more predictable memory management (#1308). However I think with that change, this trait would be incompatible because it depends on the temporary `PyBytes` extraction.

As we have `FromPyObject for OsString`, perhaps best to skip out on this trait in this PR? Can leave an issue and/or FIXME and I'm happy to revisit once #1308 is more complete.

Reply:
> However I think with that change, this trait would be incompatible because it depends on the temporary PyBytes extraction.

`&'a PyAny` would become `PyAny<'a>`, and how the lifetime works would not change?
I agree that this trait is not very necessary (since paths are often not so long and copying is not a large burden), though.